### PR TITLE
Not sure if this is correct, but implements createFromRequest

### PR DESCRIPTION
**Pull request by oliversong**
Not too sure about Mongoose stuff. Does the new instance of the model correspond to the value you would get from the db? If so can that be directly returned? That's what I did here.

### CODE REVIEW

File: app/models/operation.js [hunk 1]
```diff
@@ -77,7 +77,20 @@ var OperationSchema = mongoose.Schema({
  *   Array of Operation objects
  */
 OperationSchema.statics.createFromRequest(req) = function(request, cb) {
-
+  i=0;
+  successful = [];
+  while(i<request.length){
```
Comment (wangjohn) [inline app/models/operation.js]: Can't this just be a for loop? Might be simpler.

Comment (eob) [inline app/models/operation.js]: Agree. Consolidates termination logic into the same line as the loop command -> less risk.

File: app/models/operation.js [hunk 2]
```diff
@@ -77,7 +77,20 @@ var OperationSchema = mongoose.Schema({
...
+    var op = new Operation({ request[i] });
+    op.save(function(err){
+      if(err){
+        console.log('Did not save correctly.');
```
Comment (eob) [inline app/models/operation.js]: Shouldn't this propagate the error to the calback?

File: app/models/operation.js [hunk 3]
```diff
@@ -77,7 +77,20 @@ var OperationSchema = mongoose.Schema({
...
+      }else{
+        successful.push(op);
```
Comment (eob) [inline app/models/operation.js:96]: Shouldn't this send the result to the callback?

File: app/models/operation.js [hunk 4]
```diff
@@ -77,7 +77,20 @@ var OperationSchema = mongoose.Schema({
...
+      }
+    }
+    i++;
+  };
+  return successful;
```
Comment (eob) [inline app/models/operation.js]: Premature. successful === [] at this point. All the logic which modifies it is embedded in a closure that (in most cases) has yet to be executed at this point. No return value, unless I'm missing something.

File: app/models/operation.js (hunk 1 above)
Comment (oliversong) [inline app/models/operation.js]: I wasn't too sure if the array needed to be in the correct operation order. Unlike python, js for loops arbitrarily order the input, I think?

Comment (oliversong) [inline app/models/operation.js]: If it's not needed, I'm happy to change it back.

File: app/models/operation.js (hunk 4 above)
Comment (oliversong) [inline app/models/operation.js]: Aha, right, node.js. Will fix!

File: app/models/operation.js (hunk 1 above)
Comment (oliversong) [inline app/models/operation.js]: Ohhh, now I see what you were talking about. Man, I'm spoiled by Coffeescript. Either way, it's been replaced with async.

**Comment by eob** [comment]
Merging, but I'm going to overwrite your createFromRequest because it still returns prematurely.

**Comment by oliversong** [comment]
Wait, what? How? Oh, is it because op.save is within the loop? What is the correct way to do that?

**Comment by oliversong** [comment]
Hmm ok, I see.

**Comment by eob** [comment]
Oops!

Sorry I didn't see all these replies for some reason.

I pushed up some code to handle it w/o the async library. But yeah the key
is that you don't want to advance to the next element until you are
positive that this element is done processing.

So if you say:

obj.save(function() { success! }, function() { error! })
advance();

Then you're going to advance to the next element while the database is
still processing the save method. Instead:

obj.save(function() { success! advance(); }, function() { error! advance();
})

Async code is tricky, but once you get the hang of the style it'll feel
natural.

Sorry to smash-push changes over yours. I wanted to make sure that we had
an end-to-end system up and running again so that we're in a good place to
converge and then break back apart for next week.  Also, btw, nice w/ the
CSS -- it looks great on every site I've tried it on so far.

On Thu, Oct 17, 2013 at 12:29 PM, Oliver Song notifications@github.comwrote:

> Hmm ok, I see.
> 
> —
> Reply to this email directly or view it on GitHubhttps://github.com/cts/cts-server/pull/1#issuecomment-26524174
> .

## 

:: Ted Benson
:: http://people.csail.mit.edu/eob/
:: @edwardbenson

**Pull request merged**